### DISCUSSION TITLE
Weird unicode character  when using stdin stdout without buffer.

### DISCUSSION THREAD

**Discussion by techcatgato** [pybricks/support#1822] [Q&A]
Hi all! I "solved" my previous problem of using stdin stdout "normally" without a buffer with this code:
on hub:

```
from pybricks.pupdevices import Motor
from pybricks.parameters import Port
from pybricks.tools import wait

# Standard MicroPython modules
from usys import stdin, stdout
from uselect import poll

motor = Motor(Port.A)

# Optional: Register stdin for polling. This allows
# you to wait for incoming data without blocking.
keyboard = poll()
keyboard.register(stdin)

while True:

    # Let the remote program know we are ready for a command.
    stdout.write("rdy")

    # Optional: Check available input.
    while not keyboard.poll(0):
        # Optional: Do something here.
        wait(10)

    # Read three bytes.
    cmd = stdin.read()

    # Decide what to do based on the command.
    if cmd == "fwd":
        motor.dc(50)
    elif cmd == "rev":
        motor.dc(-50)
    elif cmd == "bye":
        break
    else:
        motor.stop()

```
on pc:

```
# SPDX-License-Identifier: MIT
# Copyright (c) 2020 Henrik Blidh
# Copyright (c) 2022-2023 The Pybricks Authors

"""
Example program for computer-to-hub communication.

Requires Pybricks firmware >= 3.3.0.
"""

import asyncio
from contextlib import suppress
from bleak import BleakScanner, BleakClient
from ustruct import unpack

PYBRICKS_COMMAND_EVENT_CHAR_UUID = "c5f50002-8280-46da-89f4-6d8051e4aeef"

# Replace this with the name of your hub if you changed
# it when installing the Pybricks firmware.
HUB_NAME = "TC"


async def main():
    main_task = asyncio.current_task()

    def handle_disconnect(_):
        print("Hub was disconnected.")

        # If the hub disconnects before this program is done,
        # cancel this program so it doesn't get stuck waiting
        # forever.
        if not main_task.done():
            main_task.cancel()

    ready_event = asyncio.Event()

    def handle_rx(_, data: bytearray):
        if data[0] == 0x01:  # "write stdout" event (0x01)
            print("hi")
            payload = data.decode()

            if payload == "rdy":
                ready_event.set()
            else:
                print("Received:", payload)

    # Do a Bluetooth scan to find the hub.
    device = await BleakScanner.find_device_by_name(HUB_NAME)

    if device is None:
        print(f"could not find hub with name: {HUB_NAME}")
        return

    # Connect to the hub.
    async with BleakClient(device, handle_disconnect) as client:

        # Subscribe to notifications from the hub.
        await client.start_notify(PYBRICKS_COMMAND_EVENT_CHAR_UUID, handle_rx)

        # Shorthand for sending some data to the hub.
        async def send(data):
            # Wait for hub to say that it is ready to receive data.
            await ready_event.wait()
            # Prepare for the next ready event.
            ready_event.clear()
            # Send the data to the hub.
            await client.write_gatt_char(
                PYBRICKS_COMMAND_EVENT_CHAR_UUID,
                b"\x06" + data,  # prepend "write stdin" command (0x06)
                response=True
            )

        # Tell user to start program on the hub.
        print("Start the program on the hub now with the button.")

        # Send a few messages to the hub.
        for i in range(5):
            await send("fwd")
            await asyncio.sleep(1)
            await send("rev")
            await asyncio.sleep(1)
            print(".", end="", flush=True)

        # Send a message to indicate stop.
        await send("bye")

        print("done.")

    # Hub disconnects here when async with block exits.


# Run the main async program.
if __name__ == "__main__":
    with suppress(asyncio.CancelledError):
        asyncio.run(main())
```

However. when i turn on this program i get:
![image](https://github.com/user-attachments/assets/28b86971-de49-4abb-8f44-51d1ae7d237d)
Any attempt so far to get rid of the weird unicode character in the beginining didnt work, and make the hub disconnect. ive tried replace(), strip(), and getting [:1] and all of these make the hub disconnect. Modifying the pc code to handle this character in the if loop also disconnects. Do you know any way of getting rid of this character?

**Comment by BertLindeman** [comment]
Have played your scenario here on a prime hub with a non stable firmware on windows 11.

Seen:

```
python issue_1822_pc_program.py
Start the program on the hub now with the button.
hi
Received: ('primehub', '3.5.0
hi
Received: ', 'ci-build-3540-v
hi
Received: 3.
hi
Received: 5.0-90-g9d3a6f85 on
hi
Received:  2024-09-06')

hi
Received: rdy
hi
Received: The program was sto
hi
Received: pped (SystemExit).
```

No garbage before `rdy`.

What firmware do you use. (I did not remember...)

Bert

**Reply by techcatgato** [reply]
Hi, where can I find the ZIP for the latest non stable version?

**Comment by BertLindeman** [comment]
In [pybricks-micropython actions](https://github.com/pybricks/pybricks-micropython/actions/workflows/build.yml)

The top entry is the latest.

**Reply by techcatgato** [reply]
Cant manage to reinstal lthe nightly version. Another thing, im using pycharm for this

**Reply by BertLindeman** [reply]
Couldn't you just use a compatible  browser on windows / linux / android just to install the firmware?

**Reply by techcatgato** [reply]
> Couldn't you just use a compatible browser on windows / linux / android just to install the firmware?

when i try to the hub isnt on the connection menu, and i have the driver

**Reply by BertLindeman** [reply]
From what environment / what browser is that?

**Reply by techcatgato** [reply]
> From what environment / what browser is that?

Chrome on a guest tab

**Reply by BertLindeman** [reply]
There might be some experimental `chrome://flags` needed to enable, but I am not sufficiently known with that.

Sorry.

Maybe [Pybricks getting started](https://pybricks.com/learn/getting-started/what-do-you-need/) can help?

Bert

**Reply by techcatgato** [reply]
> There might be some experimental `chrome://flags` needed to enable, but I am not sufficiently known with that.
> 
> Sorry.
> 
> Maybe [Pybricks getting started](https://pybricks.com/learn/getting-started/what-do-you-need/) can help?
> 
> Bert

Hi, after a lot of time I just can't install nightly. That should be fine imo cause my pybricks firmware is very new and the chances that i just missed the update that fixes that are low imo. Do you think theres anything else that can cause the weird ? every method of decoding in python i found leads to "rdy"

**Comment by techcatgato** [comment]
I have identified the character to be u\0001, But I still can't get rid of it or process it, even with the beta firmware i managed to install

**Reply by BertLindeman** [reply]
As I do not have a clue, I asked chatGPT. 

Part of the elaborate answer:

`
The u\0001 (or \x01 in hexadecimal) is the Start of Heading (SOH) control character in ASCII. 
It's a non-printable character that can appear for several reasons.
`

No idea if this would help, this is the suggested snippet to take the u\0001 off:

```python
# Strips the SOH character if present
def clean_data(data):
    if data.startswith('\x01'):
        data = data[1:]  # Remove the first character
    return data
```

```python
# Usage example
raw_data = '\x01fwd'
cleaned_data = clean_data(raw_data)
print(cleaned_data)  # Output: 'fwd'
```

Bert

**Reply by BertLindeman** [reply]
Took another look at the pybricks issues and found #1052.

Would the sample programs here [issue 1052](https://github.com/pybricks/support/issues/1052#issue-1688100446) help?

The output I get from `python issue_1052_pc_program.py`

```
06:DF:F8:A6:A8:79: Spike 3540 slot
connected!
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████| 373/373 [00:00<00:00, 5.07kB/s]
writing line...
writing line...
48 49 50 51 52 53 13 10
line received
bytearray(b'012345\r\n')
writing line...
104 101 108 108 111 13 10
line received
bytearray(b'hello\r\n')
13 10
line received
bytearray(b'\r\n')
OK!
```

Bert

**Reply by techcatgato** [reply]
Hi, this seems to be with pybricksdev no? I use regular pybricks

**Reply by BertLindeman** [reply]
The windows side uses pybricksdev, yes.

Might there be in the programs you use a statement the prefixes the data to be transmitted with 0x01?
I **have** seen an example that does that to indicate the payload is a command.
Have been searching for a long time today, but I cannot find it again. 😢 

Can you share the programs here as they are now?

**Reply by techcatgato** [reply]
The code I wrote in the beginning of this post is still the exact same and it's all I'm using, nothing more. Pc side is in pycharm

**Reply by BertLindeman** [reply]
On windows? 

I know nothing about pycharm.
These tests run in a command prompt, here.
Could you run the PC side in a simple command prompt?

**Reply by techcatgato** [reply]
I'll try. Do you mean the regular python cmd like console?

**Reply by BertLindeman** [reply]
yes

**Reply by techcatgato** [reply]
I solved this but I have another problem, now when the payload == "rdy" is true, the hub disconnects when it tries to execute ready_event_set()

**Reply by dlech** [reply]
If you can share a Python script and Pybricks firmware version that reproduces the problem, we can look into it.
